### DISCUSSION TITLE
Some important naming conventions for buckyos

### DISCUSSION THREAD

**Discussion by lurenpluto** [buckyos/buckyos#46] [Announcements]
#### Description

I've observed that our project currently has some inconsistencies in the naming conventions of parameters, constants, files, and project names. To improve the project's readability, maintainability, and to align with common Rust community practices, I propose the following guidelines:

#### Proposed Naming Conventions

1. **File Names**: Use `snake_case` (e.g., `config.rs`).
2. **Struct Fields**: Use `snake_case`.
3. **JSON Field Names**: Use `snake_case`.
4. **Struct/JSON Field Values**:
   - There are no strictly enforced rules on whether to use `snake_case` or `kebab-case`. It often depends on the specific application scenario and team preferences. For example, using `tunnel_name: "cyfs-gateway"` might be more appropriate for certain cases.
5. **Environment Variable Names**: Use `UPPER_SNAKE_CASE`.
6. **Command Line Arguments and Environment Variable Values**: Use `kebab-case`.
7. **Project Names**:
   - While there is no strict rule for naming Rust projects, it's generally more common and recommended to use `snake_case` as per Rust community practices.

#### Example

- **Current**: 
  - File Name: `Config.rs`
  - Struct Field: `tunnelName`
  - JSON Field Name: `"tunnel-name"`
  - Struct/JSON Field Value: `"cyfs_gateway"`
  - Environment Variable Name: `cyfsGateway`
  - Command Line Argument: `--tunnel_name`
  
- **Proposed**:
  - File Name: `config.rs`
  - Struct Field: `tunnel_name`
  - JSON Field Name: `"tunnel_name"`
  - Struct/JSON Field Value: `"cyfs-gateway"`
  - Environment Variable Name: `CYFS_GATEWAY`
  - Command Line Argument: `--tunnel-name`

#### Benefits

- **Readability**: Consistent naming conventions improve the readability of the code.
- **Maintainability**: Easier for developers to follow the same guidelines, resulting in less cognitive load.
- **Community Alignment**: Adhering to common Rust community conventions makes our project more welcoming to new contributors familiar with Rust.

#### Additional Recommendation: Tooling Assistance

To help maintain consistent naming conventions and code quality throughout the project, consider using the following tools:

- **`rustfmt`**: Used for automatically formatting Rust code, ensuring a consistent code style across the project.

- **`clippy`**: Useful for static code analysis, helping to identify potential issues and adhere to best practices.

- **Configuration Example**:

  To configure `rustfmt` and `clippy` according to our naming conventions, we can create or update a `rustfmt.toml` and `.clippy.toml` file in the project root directory. For example, in `rustfmt.toml`, we can set the desired options:

  ```toml
  # rustfmt.toml
  normalize_comments = true
  edition = "2018"
  ```

  In `.clippy.toml`, you can do the same for the Clippy configuration:

  ```toml
  # .clippy.toml
  warnings = ["naming_convention"]
  ```

By adopting these tools, we can automate the enforcement of our naming conventions and other code quality checks, contributing to a more robust and maintainable codebase.

Let's discuss and finalize these guidelines for a consistent and maintainable codebase.

**Comment by waterflier** [comment]
Agree!
